Add BLIP Diffusion

ayushtues opened the pull request:
This PR implements BLIP Diffusion as discussed in https://github.com/huggingface/diffusers/issues/4274

Notion for tracking progress/brainstorming : [link](https://satisfying-writer-e7e.notion.site/BLIP-Diffusion-HF-5753a09877b54e33b5abcd5eb35d7a53?pvs=4) 

## Model/Pipeline Description

BLIP diffusion (Salesforce): https://dxli94.github.io/BLIP-Diffusion-website/

BLIP diffusion enables subject-driven zero-shot image generation which is probably its best USP.

Code with pre-trained weights: https://github.com/salesforce/LAVIS/tree/main/projects/blip-diffusion
Paper: https://arxiv.org/abs/2305.14720

Abstract:
> Subject-driven text-to-image generation models create novel renditions of an input subject based on text prompts. Existing models suffer from lengthy fine-tuning and difficulties preserving the subject fidelity. To overcome these limitations, we introduce BLIP-Diffusion, a new subject-driven image generation model that supports multimodal control which consumes inputs of subject images and text prompts. Unlike other subject-driven generation models, BLIP-Diffusion introduces a new multimodal encoder which is pre-trained to provide subject representation. We first pre-train the multimodal encoder following BLIP-2 to produce visual representation aligned with the text. Then we design a subject representation learning task, called prompted context generation, which enables a diffusion model to leverage such visual representation and generates new subject renditions. Compared with previous methods such as DreamBooth, our model enables zero-shot subject-driven generation, and efficient fine-tuning for customized subject with up to 20x speedup. We also demonstrate that BLIP-Diffusion can be flexibly combined with existing techniques such as ControlNet and prompt-to-prompt to enable novel subject-driven generation and editing applications.

From the official website mentioned above:
![image](https://github.com/huggingface/diffusers/assets/43698245/8e617533-f5e6-4555-8ab8-3eb9f2bf5b1e)

## TODO

- [x] Implement BLIPDiffusionPipeline
- [x] Script to convert pretrained weights into diffusers checkpoints
- [ ] Add model cards for checkpoints and move checkpoints to appropriate repositories
- [x] Write tests
- [x] Add docstrings for new classes
- [x] Create documentation 
- [x] Add usage example(s)

HF Model Link : https://huggingface.co/ayushtues/blipdiffusion/tree/main


## Usage Examples

### Zero-Shot Subject Driven Generation
```python
from diffusers.pipelines import BlipDiffusionPipeline
from diffusers.utils import load_image


blip_diffusion_pipe= BlipDiffusionPipeline.from_pretrained('ayushtues/blipdiffusion')
blip_diffusion_pipe.to('cuda')

cond_subject = ["dog"]
tgt_subject = ["dog"]
text_prompt_input = ["swimming underwater"]


cond_image = load_image("https://huggingface.co/datasets/ayushtues/blipdiffusion_images/resolve/main/dog.jpg")
num_output = 1

iter_seed = 88888
guidance_scale = 7.5
num_inference_steps = 50
negative_prompt = "over-exposure, under-exposure, saturated, duplicate, out of frame, lowres, cropped, worst quality, low quality, jpeg artifacts, morbid, mutilated, out of frame, ugly, bad anatomy, bad proportions, deformed, blurry, duplicate"

for i in range(num_output):
    output = blip_diffusion_pipe(
        text_prompt_input,
        cond_image,
        cond_subject,
        tgt_subject,
        guidance_scale=guidance_scale,
        num_inference_steps=num_inference_steps,
        neg_prompt=negative_prompt,
        height=512,
        width=512,
    )

```
Input
![image](https://github.com/huggingface/diffusers/assets/43698245/30f5af2a-3f94-4fef-a55d-0dad63e1f2e2)
Output
![image](https://github.com/huggingface/diffusers/assets/43698245/21d3da5f-ec8f-4895-a738-6cc2bc8e2327)

### Controlled subject-driven generation Canny-edge
```python
from diffusers.pipelines import BlipDiffusionControlNetPipeline
from diffusers.utils import load_image
from controlnet_aux import CannyDetector

blip_diffusion_pipe= BlipDiffusionControlNetPipeline.from_pretrained("ayushtues/blipdiffusion-controlnet")
blip_diffusion_pipe.to('cuda')

style_subject = ["flower"] # subject that defines the style
tgt_subject = ["teapot"]  # subject to generate.
text_prompt = ["on a marble table"]
cldm_cond_image = load_image("https://huggingface.co/datasets/ayushtues/blipdiffusion_images/resolve/main/kettle.jpg").resize((512, 512))
canny = CannyDetector()
cldm_cond_image = canny(cldm_cond_image, 30, 70, output_type='pil')
cldm_cond_image = [cldm_cond_image ]

style_image = load_image("https://huggingface.co/datasets/ayushtues/blipdiffusion_images/resolve/main/flower.jpg")


num_output = 1
iter_seed = 88888
guidance_scale = 7.5
num_inference_steps = 50
negative_prompt = "over-exposure, under-exposure, saturated, duplicate, out of frame, lowres, cropped, worst quality, low quality, jpeg artifacts, morbid, mutilated, out of frame, ugly, bad anatomy, bad proportions, deformed, blurry, duplicate"

for i in range(num_output):
    output = blip_diffusion_pipe(
        text_prompt,
        style_image,
         cldm_cond_image,
        style_subject,
        tgt_subject,
        guidance_scale=guidance_scale,
        num_inference_steps=num_inference_steps,
        neg_prompt=negative_prompt,
        height=512,
        width=512,
    )
```
Canny edge based Controlnet example - 
Input 
![image](https://github.com/huggingface/diffusers/assets/43698245/66391e2a-e8fb-4183-a530-b6c28bd0185a)
Conditioning image for Canny Edge 
![image](https://github.com/huggingface/diffusers/assets/43698245/f38444fa-b42b-4f52-ae57-1a79e735f5fa)
Output 
![image](https://github.com/huggingface/diffusers/assets/43698245/076486e5-d134-4710-9993-2eaf7bf30f4c)


### Controlled subject-driven generation Scribble
```python
from diffusers.pipelines import BlipDiffusionControlNetPipeline
from diffusers.utils import load_image
from controlnet_aux import HEDdetector

blip_diffusion_pipe= BlipDiffusionControlNetPipeline.from_pretrained("ayushtues/blipdiffusion-controlnet")
controlnet = ControlNetModel.from_pretrained("lllyasviel/sd-controlnet-scribble")
blip_diffusion_pipe.controlnet = controlnet
blip_diffusion_pipe.to('cuda')

style_subject = ["flower"] # subject that defines the style
tgt_subject = ["bag"]  # subject to generate.
text_prompt = ["on a table"]
cldm_cond_image = load_image("https://huggingface.co/lllyasviel/sd-controlnet-scribble/resolve/main/images/bag.png" ).resize((512, 512))
hed = HEDdetector.from_pretrained("lllyasviel/Annotators")
cldm_cond_image = hed(cldm_cond_image)
cldm_cond_image = [cldm_cond_image ]

style_image = load_image("https://huggingface.co/datasets/ayushtues/blipdiffusion_images/resolve/main/flower.jpg")


num_output = 1
iter_seed = 88888
guidance_scale = 7.5
num_inference_steps = 50
negative_prompt = "over-exposure, under-exposure, saturated, duplicate, out of frame, lowres, cropped, worst quality, low quality, jpeg artifacts, morbid, mutilated, out of frame, ugly, bad anatomy, bad proportions, deformed, blurry, duplicate"

for i in range(num_output):
    output = blip_diffusion_pipe(
        text_prompt,
        style_image,
         cldm_cond_image,
        style_subject,
        tgt_subject,
        guidance_scale=guidance_scale,
        num_inference_steps=num_inference_steps,
        neg_prompt=negative_prompt,
        height=512,
        width=512,
    )
```



Scribble example - 
Input 
![image](https://github.com/huggingface/diffusers/assets/43698245/66391e2a-e8fb-4183-a530-b6c28bd0185a)
Conditioning image for Scribble
![image](https://github.com/huggingface/diffusers/assets/43698245/8570c8f5-c4a7-4724-97b1-ebe433ed0071)
Output 
![image](https://github.com/huggingface/diffusers/assets/43698245/ff703d6a-fb90-4662-b992-cc9f8bfeff9d)



## CC
@sayakpaul

oumad commented:
It has been 2 months since this was out, I can't believe how almost no one mentions it, let alone implement it.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_4388). All of your documentation changes will be reflected on that endpoint.

ayushtues commented:
Ported the ViT visual encoder checkpoints used: https://huggingface.co/ayushtues/blipdiffusion/tree/main/vision_encoder
```python
from src.diffusers.pipelines.blip_diffusion.modeling_blip2 import Blip2VisionConfig, Blip2VisionModel

image_input_dummy = torch.zeros(1, 3, 224, 224)
visual_encoder = Blip2VisionModel.from_pretrained('ayushtues/blipdiffusion', subfolder='visual_encoder')
image_embed = visual_encoder(image_input_dummy, return_dict=True).last_hidden_state
```

NielsRogge commented:
@ayushtues maybe you can directly use BLIP-2 in Transformers as a dependency, rather than reimplementing it in Diffusers (similar to how CLIP or T5 aren't reimplemented in diffusers for Stable Diffusion)?

You can also do `from transformers.models.blip_2.modeling_blip2 import Blip2VisionModel` for instance, in case you only need the vision encoder

ayushtues commented:
Hey @NielsRogge I originally intended to do that, but as mentioned in https://github.com/huggingface/transformers/issues/25245 the implementation of Blip2 in Transformers didn't support multimodal feature extraction, so I went ahead to do a local implementation in diffusers.

If the feature gets added in the transformers implementation, we can shift to directly importing it

ayushtues commented:
Update, was able to port the model to diffusers, although a lot of the code needs refactoring/reusing and better integration

Colab link : https://colab.research.google.com/drive/1PDlO8-1kPnhTUOmQBv5a2cIBTdYp_7Pi?usp=sharing
HF Model link : https://huggingface.co/ayushtues/blipdiffusion/tree/main

cond_subject = "dog"
tgt_subject = "dog"
text_prompt_input = "swimming underwater"

Input image - 
![OIP](https://github.com/huggingface/diffusers/assets/43698245/629643cc-76d9-4473-a728-be11893c6971)

Output image - 
![download](https://github.com/huggingface/diffusers/assets/43698245/2c09974c-b740-4ba9-8b3b-368c9e52bff2)

ayushtues commented:
Hey @sayakpaul can you please do an early review of this PR?

TheOneTrueGuy commented:
This looks like great work & thanks for all you've done.
I'm looking to sample/beam search captions with BLIP or BLIP2, does this diffusers version support that too?

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/conversion_scripts.py:
```diff
@@ -0,0 +1,173 @@
+from diffusers import (
```
nit: Should go to `scripts/` and needs a better naming.

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/conversion_scripts.py:
```diff
@@ -0,0 +1,173 @@
+from LAVIS.lavis.models import load_model_and_preprocess
+import os
+import torch
+os.environ['TRANSFORMERS_CACHE'] = 'E:/diffusers/cache'
```
Not needed.

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/conversion_scripts.py:
```diff
@@ -0,0 +1,173 @@
+from transformers.models.blip_2.configuration_blip_2 import Blip2Config, Blip2Config, Blip2VisionConfig, Blip2QFormerConfig
+from src.diffusers.pipelines import BlipDiffusionPipeline
+from clip_vit_sample import create_clip_vit_L
+from LAVIS.lavis.models import load_model_and_preprocess
```
Hmm when we move this script to `scripts` I think we could add example usage in the comments and make it clear that `LAVIS` needs to be installed for running the script.

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/modeling_blip2.py:
```diff
@@ -0,0 +1,655 @@
+# coding=utf-8
+# Copyright 2023 The Salesforce Authors and The HuggingFace Team. All rights reserved.
```
BLIP Diffusion authors.

sayakpaul replied:
I think we only need to include that in `pipeline_blip_diffusion.py`.

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,431 @@
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+import re
+
+def numpy_to_pil(images):
+    """
+    Convert a numpy image or a batch of images to a PIL image.
+    """
+    if images.ndim == 3:
+        images = images[None, ...]
+    images = (images * 255).round().astype("uint8")
+    pil_images = [Image.fromarray(image) for image in images]
+
+    return pil_images
```
No need to implement this here. We already have it: https://huggingface.co/docs/diffusers/main/en/api/utilities#diffusers.utils.numpy_to_pil

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,431 @@
+    def __init__(self, tokenizer: CLIPTokenizer, text_encoder: CtxCLIPTextModel, vae: AutoencoderKL, unet: UNet2DConditionModel, scheduler: PNDMScheduler, qformer: Blip2QFormerModel):
+        super().__init__()
+
+        self._CTX_BEGIN_POS = 2
```
Could we register this as a config of the pipeline?

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,431 @@
+
+        self.register_modules(tokenizer=tokenizer, text_encoder=text_encoder,  vae=vae, unet=unet, scheduler=scheduler, qformer=qformer)
+    
+    def enable_sequential_cpu_offload(self, gpu_id=0):
+        r"""
+        Offloads all models to CPU using accelerate, significantly reducing memory usage. When called, the pipeline's
+        models have their state dicts saved to CPU and then are moved to a `torch.device('meta') and loaded to GPU only
+        when their specific submodule has its `forward` method called.
+        """
+        if is_accelerate_available():
+            from accelerate import cpu_offload
+        else:
+            raise ImportError("Please install accelerate via `pip install accelerate`")
+
+        device = torch.device(f"cuda:{gpu_id}")
+
+        models = [
+            self.text_encoder,
+            self.vae,
+            self.unet,
+            self.qformer,
+        ]
+        for cpu_offloaded_model in models:
+            if cpu_offloaded_model is not None:
+                cpu_offload(cpu_offloaded_model, device)
```
No need to implement this since `DiffusionPipeline` already has it. 

```suggestion
```

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,431 @@
+        # We'll offload the last model manually.
+        self.final_offload_hook = hook
+
+    @property
+    # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline._execution_device
+    def _execution_device(self):
+        r"""
+        Returns the device on which the pipeline's models will be executed. After calling
+        `pipeline.enable_sequential_cpu_offload()` the execution device can only be inferred from Accelerate's module
+        hooks.
+        """
+        if not hasattr(self.unet, "_hf_hook"):
+            return self.device
+        for module in self.unet.modules():
+            if (
+                hasattr(module, "_hf_hook")
+                and hasattr(module._hf_hook, "execution_device")
+                and module._hf_hook.execution_device is not None
+            ):
+                return torch.device(module._hf_hook.execution_device)
+        return self.device
```
Same. We don't need it as it's now a part of `DiffusionPipeline`.

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,431 @@
+        return self.device
+
+
+    def _build_prompt(self, prompts, tgt_subjects, prompt_strength=1.0, prompt_reps=20):
+        rv = []
+        for prompt, tgt_subject in zip(prompts, tgt_subjects):
+            prompt = f"a {tgt_subject} {prompt.strip()}"
+            # a trick to amplify the prompt
+            rv.append(", ".join([prompt] * int(prompt_strength * prompt_reps)))
+
+        return rv
```
Is this is coming from the original codebase? In any case, having a bit of a description to accompany the method would be nice.

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,431 @@
+
+        return rv
+
+    def _predict_noise(
+        self,
+        t,
+        latent_model_input,
+        text_embeddings,
+        width=512,
+        height=512,
+        cond_image=None,
+    ):
+ 
+        down_block_res_samples, mid_block_res_sample = None, None
+
+        noise_pred = self.unet(
+            latent_model_input,
+            timestep=t,
+            encoder_hidden_states=text_embeddings,
+            down_block_additional_residuals=down_block_res_samples,
+            mid_block_additional_residual=mid_block_res_sample,
+        )["sample"]
```
We always make it a part of the `__call__()`. So, let's move it to the pipeline call method.

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,431 @@
+
+        return noise_pred
+
+    def _init_latent(self, latent, height, width, generator, batch_size):
+        if latent is None:
+            latent = torch.randn(
+                (1, self.unet.in_channels, height // 8, width // 8),
+                generator=generator,
+                device=generator.device,
+            )
+        latent = latent.expand(
+            batch_size,
+            self.unet.in_channels,
+            height // 8,
+            width // 8,
+        )
+        return latent.to(self.device)
```
We make use of 

https://github.com/huggingface/diffusers/blob/078df46bc9a99178a9a744b872899990353769a4/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L502

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,431 @@
+        )
+        return latent.to(self.device)
+
+    def _forward_prompt_embeddings(self, input_image, src_subject, prompt):
+        # 1. extract BLIP query features and proj to text space -> (bs, 32, 768)
+        query_embeds = self.forward_ctx_embeddings(input_image, src_subject)
+
+        # 2. embeddings for prompt, with query_embeds as context
+        tokenized_prompt = self._tokenize_text(prompt).to(self.device)
+        text_embeddings = self.text_encoder(
+            input_ids=tokenized_prompt.input_ids,
+            ctx_embeddings=query_embeds,
+            ctx_begin_pos=[self._CTX_BEGIN_POS],
+        )[0]
+
+        return text_embeddings
```
I think there are two things in here. WDYT about having an `encode_prompt()` method and  a separate `get_query_embeddings()` method? 

This way, we would promote a better separation of concern and composability. 

This will also allow the users to explicitly pass the `query_embeds` for whatever reason (memory savings, etc.). 

Or we, can directly make use of `query_embeds = self.forward_ctx_embeddings(input_image, src_subject)` and propagate `query_embeds` to `encode_prompt()` subsequently. 

Does this make sense?

ayushtues replied:
I think it makes sense to have a separate `get_query_embeddings` and `encode_prompt`, to enable getting the query embeds if necessary, caching them and then passing to `encode_prompt`, if something like that is required.

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,431 @@
+            encoding = encoding_dist.sample(generator=rng_generator)
+        else:
+            encoding = encoding_dist.mode()
+        latents = encoding * 0.18215
```
`0.18215` should be coming from a config of the VAE. 

An example:

https://github.com/huggingface/diffusers/blob/078df46bc9a99178a9a744b872899990353769a4/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L845

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,431 @@
+
+        return caption
+    
+    def preprocess_image(self, image):
+                
+        mean = (0.48145466, 0.4578275, 0.40821073)
+        std = (0.26862954, 0.26130258, 0.27577711)
+        transform = transforms.Compose(
+            [
+                transforms.Resize(224, interpolation=InterpolationMode.BICUBIC),
+                transforms.CenterCrop(224),
+                transforms.ToTensor(),
+                transforms.Normalize(mean, std)
+            ]
+        )
+
+        return transform(image)
```
We should not introduce a dependency on `torchvision` here and instead adapt: 

https://github.com/huggingface/diffusers/blob/078df46bc9a99178a9a744b872899990353769a4/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_instruct_pix2pix.py#L296

Cc: @yiyixuxu for refactoring needed to support this preprocessing.

sayakpaul replied:
Thanks for explaining. 

Yeah let's do that in the image_processor.py module.

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,431 @@
+    @torch.no_grad()
+    def __call__(
+        self,
+        samples,
```
Let's spit this out in the spirit of utmost clarity. 

We could do:

* `prompt`
* `reference_image`
* `source_subject_category`
* `target_subject_category`

We then pass all of these to `check_inputs` to validate if their shapes and sizes conform to what we're expecting. 

All of these could either be single values or could be lists similar to how it's done in the other pipelines.

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,431 @@
+
+        return image
+
+    def _latent_to_image(self, latents):
+        latents = 1 / 0.18215 * latents
+        image = self.vae.decode(latents).sample
+        print(torch.mean(image))
+        image = (image / 2 + 0.5).clamp(0, 1)
+        image = image.cpu().permute(0, 2, 3, 1).numpy()
+
+        image = numpy_to_pil(image)
+
+        return image
```
We should make use of https://github.com/huggingface/diffusers/blob/078df46bc9a99178a9a744b872899990353769a4/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_instruct_pix2pix.py#L398

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,431 @@
+
+        return image
+
+    def _denoise_latent_step(
```
This should be a part of `__call()__` similar to the rest of the pipelines.

sayakpaul reviewed (commented):
Thanks so much for your hard work!

Left some comments on the design. 

If possible, could you enlist the possible use cases enabled by the current PR? IIRC, BLIP Diffusion supports multiple use cases, right?

ayushtues commented:
Thanks @sayakpaul for the detailed review, will integrate the suggestions soon!

ayushtues commented:
Hi @sayakpaul so Blip-Diffusion has 4 major usecases

1. Zero-Shot Subject Driven Generation - Generate the image of a given subject in a new scene, from a single image. This is the current PR
2. Few Step Finetuned Subject Driven Generation - Fine-tune the model with 3-4 images of the subject. And generate new images of it with a better quality
3. Controlled subject-driven generation - Controlnet + BlipDiffusion
4. Subject-driven editing via cross-attention control - Prompt2Prompt + BlipDiffusion

ayushtues commented:
This PR supports only 1 as of now.

sayakpaul commented:
> so Blip-Diffusion has 4 major usecases

Yup. 

I am okay to start with 1 as that is still very useful! 

> Few Step Finetuned Subject Driven Generation - Fine-tune the model with 3-4 images of the subject. And generate new images of it with a better quality

I guess this requires us to add a fine-tuning script. We can leave it up to the community to pick up. If you want to add it at a later point in time, more than happy to help with that!

> Controlled subject-driven generation - Controlnet + BlipDiffusion

I would assume we can support this readily, no? Since the base UNet and the VAE are based on existing SD models.

> Subject-driven editing via cross-attention control - Prompt2Prompt + BlipDiffusion

Okay for me to ignore for now. 

Let me know what you think.

ayushtues commented:
Yeah makes sense, let me try to add ControlNet support as well in this PR

sayakpaul commented:
@ayushtues let me know whenever you need any help. Keen on getting this shipped!

ayushtues commented:
Added Controlnet, colab link : https://colab.research.google.com/drive/1PTWI_NUM2WAz9wfAzbnNxQuo89a_uCa7?usp=sharing

Canny edge example : 
![image](https://github.com/huggingface/diffusers/assets/43698245/d0dbc6c7-92e7-4f81-9dc9-44c096b45861)
![image](https://github.com/huggingface/diffusers/assets/43698245/3bc62b2e-7f8c-468a-9b5d-2a83e25d6bb7)
![image](https://github.com/huggingface/diffusers/assets/43698245/dcd7d1b4-9c5a-457e-a25d-be9900c2368f)

A lot of extra-preprocessing for canny edge detection is happening outside the pipeline (as can be seen in the colab), needs some refactoring.

sayakpaul commented:
Oh wow! Let me know if you would like a review!

ayushtues commented:
Hi @sayakpaul, I moved the processor to a `BlipImageProcessor` https://github.com/ayushtues/diffusers/blob/05451f3ea22470322034297360851fb8de357264/src/diffusers/image_processor.py#L369, maybe you can review it once? Also, the ControlNet functionality.

Next steps are writing some tests, and adding the ControlNet processing to `BlipImageProcessor`

ayushtues commented:
Added a fast test

NielsRogge commented on src/diffusers/pipelines/blip_diffusion/modeling_blip2.py:
```diff
@@ -0,0 +1,653 @@
+# coding=utf-8
```
Would you be able to contribute the get_image `features` and `get_text_features` methods to the Transformers library, such that this file can be omitted?

sayakpaul replied:
For now, I think it's okay for the PR as it doesn't get blocked.

sayakpaul commented on scripts/convert_blipdiffusion_to_diffusers.py:
```diff
@@ -0,0 +1,169 @@
+from src.diffusers.pipelines.blip_diffusion.modeling_ctx_clip import CtxCLIPTextModel
+from transformers.models.blip_2.configuration_blip_2 import Blip2Config, Blip2Config, Blip2VisionConfig, Blip2QFormerConfig
+from src.diffusers.pipelines import BlipDiffusionPipeline
+from LAVIS.lavis.models import load_model_and_preprocess
```
We should either add a utility here (`is_lavis_available()`) or make it clear from the script that `LAVIS` is expected to load this model. See how it's done here: 

https://github.com/huggingface/diffusers/blob/d6141205cd69f3ad872e5a905af0b92d5f03709f/scripts/convert_original_audioldm2_to_diffusers.py#L48

ayushtues replied:
We need to build lavis from source for this, even the version in pip doesn't have blip-diffusion currently. Added comment for this in the code

sayakpaul commented on scripts/convert_blipdiffusion_to_diffusers.py:
```diff
@@ -0,0 +1,169 @@
+    skip_prk_steps=True,
+)
+
+#TODO: Test this once
+tokenizer = CLIPTokenizer.from_pretrained(
+    "./runwayml/stable-diffusion-v1-5/", subfolder="tokenizer", cache_dir='./cache'
+)
+blipDiffusion = BlipDiffusionPipeline(tokenizer=tokenizer, text_encoder=text_encoder,  vae=vae, unet=unet, scheduler=scheduler, qformer=qformer)
+blipDiffusion.save_pretrained("blip_diffusion")
```
Could we make it a nice function following the structure of the rest of the scripts present inside `scripts/`?

ayushtues replied:
Done, some funny stuff is happening an there seems to be a difference in outputs now

sayakpaul commented on src/diffusers/image_processor.py:
```diff
@@ -364,3 +364,272 @@ def postprocess(
+
+class BlipImageProcessor(ConfigMixin):
+    """
+    Image processor for VAE.
```
Needs to change.

sayakpaul replied:
Could we not reuse https://huggingface.co/docs/transformers/main/en/model_doc/blip#transformers.BlipImageProcessor ? Let's outline the differences from what's provided in `transformers` and see how we can reuse it as much as possible.

ayushtues replied:
We can maybe create a new processor, which has a `BlipImageProcessor` object, and use it and add center_crop on top of it

sayakpaul replied:
> We can maybe create a new processor, which has a BlipImageProcessor object, and use it and add center_crop on top of it

To me this sounds very good. 

Ccing @yiyixuxu for the best way forward here. 

Maybe also cc @amyeroberts to check if she has any other inputs to reuse things from `transformers` here.

amyeroberts replied:
There was some recent refactoring of the image processors, and now all image processors have `center_crop` as a method (which uses the functionality in the `image_transforms` module). So you should be able to use / copy the BlipImageProcessor directly :)

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/modeling_blip2.py:
```diff
@@ -0,0 +1,653 @@
+# coding=utf-8
+# Licensed under the Apache License, Version 2.0 (the "License");
```
We're missing :

```
# Copyright 2023 The HuggingFace Team. All rights reserved.
```

ayushtues replied:
Done

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/modeling_blip2.py:
```diff
@@ -0,0 +1,653 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" PyTorch BLIP-2 model."""
```
Maybe make it clear that:

* This exists because there's currently not support for getting multimodal embeddings. 
* There's a plan on getting that merged in `transformers` and when it's in this module will go away and instead, we're directly leverage it from `transformers`.

ayushtues replied:
Done

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/modeling_ctx_clip.py:
```diff
@@ -0,0 +1,217 @@
+"""
+ Copyright (c) 2023, salesforce.com, inc.
+ All rights reserved.
+ SPDX-License-Identifier: BSD-3-Clause
+ For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+"""
```
We have a specific format for authoring license headers. Let's follow that.

ayushtues replied:
Done

sayakpaul replied:
It's not reflecting in the commits somehow.

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/modeling_ctx_clip.py:
```diff
@@ -0,0 +1,217 @@
+)
+
+
+class CtxCLIPTextModel(CLIPPreTrainedModel):
```
Maybe spell out what `Ctx` is? I think that would be a better name.

ayushtues replied:
I don't think it has a name, I checked the BLIP paper as well, maybe call it `context`?

sayakpaul replied:
`ContextCLIPTextModel` sounds better to me!

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/modeling_ctx_clip.py:
```diff
@@ -0,0 +1,217 @@
+        return mask
+
+
+class CtxCLIPTextEmbeddings(nn.Module):
```
Could benefit from brief docstrings.

ayushtues replied:
Added

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,337 @@
+        else:
+            # image batch size is the same as prompt batch size
+            # repeat_by = num_images_per_prompt
+            raise NotImplementedError
```
Maybe raise this error with a better error message?

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,337 @@
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+import re
+
+def prepare_cond_image(
```
Is it not possible to use the image processor module here?

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,337 @@
+    def __init__(self, tokenizer: CLIPTokenizer, text_encoder: CtxCLIPTextModel, vae: AutoencoderKL, unet: UNet2DConditionModel, scheduler: PNDMScheduler, qformer: Blip2QFormerModel, controlnet: ControlNetModel=None, ctx_begin_pos: int = 2):
+        super().__init__()
+
+        self.image_processor = BlipImageProcessor()
```
The `image_processor` also needs to be registered.

sayakpaul replied:
Does `ControlNet` always need to be present to use this pipeline? I think it'd be better to create two pipelines here:

* `BlipDiffusionPipeline`
* BlipDiffusionControlNetPipeline`

ayushtues replied:
Done, created a new `BlipDiffusionControlNetPipeline`

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,337 @@
+        source_subject_category = [self.preprocess_caption(source_subject_category)]
+        target_subject_category = [self.preprocess_caption(target_subject_category)]
+
+        mean = (0.48145466, 0.4578275, 0.40821073)
+        std = (0.26862954, 0.26130258, 0.27577711)
+
```
These two should be configuration variables.

ayushtues replied:
Done

sayakpaul reviewed (commented):
Thanks a lot for the ControlNet addition! Did another pass and left some comments. I think your structure is already in good shape and I left a couple of design-related food for thought. 

Would be also great to have some ControlNet examples in the PR thread.

sayakpaul commented on scripts/convert_blipdiffusion_to_diffusers.py:
```diff
@@ -0,0 +1,219 @@
+    tokenizer = CLIPTokenizer.from_pretrained(
+        "runwayml/stable-diffusion-v1-5", subfolder="tokenizer"
+    )
+    blipDiffusion = BlipDiffusionPipeline(tokenizer=tokenizer, text_encoder=text_encoder,  vae=vae, unet=unet, scheduler=scheduler, qformer=qformer)
```
No camel casing please.

ayushtues replied:
Done

sayakpaul commented on tests/pipelines/blipdiffusion/test_blipdiffusion.py:
```diff
@@ -0,0 +1,195 @@
+
+        assert image.shape == (64, 64, 4)
+
+        expected_slice = np.array(
+            [0.5686274509803921, 0.5176470588235295, 0.6470588235294118, 0.44313725490196076, 0.6352941176470588, 0.4980392156862745, 0.396078431372549, 0.5529411764705883, 0.5411764705882353]
+        )
```
Please keep these to four precisions only.

ayushtues replied:
Done

sayakpaul reviewed (commented):
Thanks for the changes! 

I guess we're missing the test suite for the ControlNet pipeline? 

Additionally, it would be very nice to update the description of the PR with example code snippets covering various use cases and their outputs.

ayushtues commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,334 @@
+# Copyright 2023 The HuggingFace Team. All rights reserved.
```
Done

ayushtues commented:
Hi @sayakpaul can you review this again? Added some examples, docstrings, and test for ControlNet. Also copied the BlipImageProcessor from Transformers, and modified it for all the preprocessing

sayakpaul commented:
@ayushtues could you make the example snippets full-fledged please? Currently it seems like you're loading the images locally.

sayakpaul commented on scripts/convert_blipdiffusion_to_diffusers.py:
```diff
@@ -0,0 +1,221 @@
+'''
+This scripts requires you to build lavis from source, since the pip version doesn't have blip diffusion
+Follow instruction https://github.com/salesforce/LAVIS/tree/main
+'''
```
```suggestion
'''
This script requires you to build `LAVIS` from source, since the pip version doesn't have BLIP Diffusion. Follow instructions here: https://github.com/salesforce/LAVIS/tree/main.
'''
```

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/blip_image_processing.py:
```diff
@@ -0,0 +1,454 @@
+
+
+
+# Copied from transformers.models.blip.image_processing_blip.BlipImageProcessor
```
Why are we having to copy it? What differs here?

ayushtues replied:
I modified it to use `center_crop` as well which was not there in the original code. Also extended it to add a new `canny_preprocess` function to process the canny edge images

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/modeling_blip2.py:
```diff
@@ -0,0 +1,657 @@
+
+logger = logging.get_logger(__name__)
+
+""" PyTorch BLIP-2 model.
+    There is an implementation of Blip2 in Transformers : https://github.com/huggingface/transformers/blob/main/src/transformers/models/blip_2/modeling_blip_2.py
+    Which doesn't support getting multimodal embeddings for now
+    Will replace this with the Transformers version once it supports multimodal embeddings 
+"""
```
```suggestion
# There is an implementation of Blip2 in `transformers` : https://github.com/huggingface/transformers/blob/main/src/transformers/models/blip_2/modeling_blip_2.py. 
# But it doesn't support getting multimodal embeddings. So, this module can be
# replaced with a future `transformers` version supports that. 
```

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/modeling_blip2.py:
```diff
@@ -0,0 +1,657 @@
+
+
+
+# Copied from transformers.models.blip.modeling_blip.BlipVisionEmbeddings with Blip->Blip2
```
This "Copied from ...` comments (applicable for the comments that have `transformers` mentions) will break our CI since there's no local module named `transformers`. So, maybe let's just stick to `# Copy-pasted ...`.

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/modeling_blip2.py:
```diff
@@ -0,0 +1,657 @@
+
+        self.class_embedding = nn.Parameter(torch.randn(1, 1, self.embed_dim))
+
+        # TO-DO Added bias = False here, not in OG HF code
```
Explain this comment?

ayushtues replied:
I just added it to keep track of the differences with the implementation of Blip2 in Transfromers and the one in LAVIS. Removing it

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/modeling_blip2.py:
```diff
@@ -0,0 +1,657 @@
+        embeddings = embeddings + self.position_embedding[:, : embeddings.size(1), :].to(target_dtype)
+        return embeddings
+
+class Blip2QFormerEncoder(nn.Module):
```
Brief docstrings.

ayushtues replied:
Done!

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/modeling_blip2.py:
```diff
@@ -0,0 +1,657 @@
+
+
+
+class Blip2QFormerLayer(nn.Module):
```
Same as above.

ayushtues replied:
Done!

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/modeling_ctx_clip.py:
```diff
@@ -0,0 +1,229 @@
+    _expand_mask,
+)
+
+'''
+This is a modified version of the CLIPTextModel from transformers.models.clip.modeling_clip
+Which allows for an extra input of "context embeddings", which are the query embeddings used in Qformer
+They pass through the clip model, along with the text embeddings, and interact with them using self attention
+'''
```
Let's turn this into `#` style comments.

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,383 @@
+# Copyright 2023 Blip Diffusion authors and The HuggingFace Team. All rights reserved.
```
Let's stay consistent. In one place we write "Salesforce" and here we are using "Blip Diffusion". Let's maybe go with Salesforce.

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,383 @@
+        >>> from diffusers import BlipDiffusionPipeline
+        >>> from PIL import Image
+
+        >>> blipDiffusion = BlipDiffusionPipeline.from_pretrained('ayushtues/blipdiffusion')
```
`blip_diffusion_pipe` please.

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
````diff
@@ -0,0 +1,383 @@
+        ```
+"""
+
+# Create a class for the Blip Diffusion pipeline
````
I don't think this comment is needed.

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,383 @@
+            Standard deviation of the image processor.
+    """
+    
+    def __init__(self, tokenizer: CLIPTokenizer, text_encoder: ContextCLIPTextModel, vae: AutoencoderKL, unet: UNet2DConditionModel, scheduler: PNDMScheduler, qformer: Blip2QFormerModel,  image_processor: BlipImageProcessor, ctx_begin_pos: int = 2, mean: tuple = (0.48145466, 0.4578275, 0.40821073), std: tuple = (0.26862954, 0.26130258, 0.27577711)):
```
Configuration works for me!

ayushtues replied:
I wasn't sure what you meant, can you check if the current thing is correct?

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,383 @@
+
+        return rv
+
+    def prepare_latents(self, batch_size, num_channels, height, width, dtype, device, generator, latents=None):
```
Let's make sure to add "# Copied from ..." if that's been the case.

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,383 @@
+    @replace_example_docstring(EXAMPLE_DOC_STRING)
+    def __call__(
+        self,
+        prompt,
+        reference_image,
+        source_subject_category,
+        target_subject_category,
+        latents=None,
+        guidance_scale=7.5,
+        height=512,
+        width=512,
+        seed=42,
+        num_inference_steps=50,
+        neg_prompt="",
+        prompt_strength=1.0,
+        prompt_reps=20,
```
Type annotations would be amazing here.

ayushtues replied:
Done!

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,383 @@
+            `List[PIL.Image.Image]` : The generated images.
+        """
+
+        prompt = [self.preprocess_caption(prompt)]
+        source_subject_category = [self.preprocess_caption(source_subject_category)]
+        target_subject_category = [self.preprocess_caption(target_subject_category)]
```
Why do we need to additionally preprocess here? I mean we don't do this for the rest of the pipelines. Could I get a brief explanation behind doing this?

ayushtues replied:
I added it to replicate the processing done in the original paper. On a closer look, it seems to just clean the prompt using basic regex, we can remove it if needed

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,383 @@
+        text_embeddings = self.encode_prompt(
+            query_embeds, prompt
+        )
+        # 3. unconditional embedding
```
There's nothing for # 1 and # 2.

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,383 @@
+                    noise_pred_text - noise_pred_uncond
+                )
+
+            # TODO - Handle ddim as well
+            # # compute the previous noisy sample x_t -> x_t-1
+            # scheduler = self.ddim_scheduler if use_inversion else self.pndm_scheduler
```
Briefly explain?

ayushtues replied:
They used DDIM to do inversion for editing real images, kept a note here for future, removing for now

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,383 @@
+        if with_query:
+            max_len -= self.qformer.config.num_query_tokens
+
+        tokenized_text = self.tokenizer(
```
We usually keep it as a part of the `encode_prompt()` method itself.

sayakpaul reviewed (commented):
Kudos for your hard work, Ayush! Looking fire. Left another round of comments. 

Let's maybe start working on fixing the failures from here on to ship this soon!

ayushtues commented:
> @ayushtues could you make the example snippets full-fledged please? Currently it seems like you're loading the images locally.

Done!

ayushtues commented:
@sayakpaul seems like the original Blip Diffusion doesn't support batch_size > 1, do we need to support batched inference?

sayakpaul commented:
Would be great if we want to maximize usage.

ayushtues commented:
Okay, will need some changes, let me give it a shot

ayushtues commented:
Hi @sayakpaul I have added batch support, and all tests pass on my local machine now. Can you do a review once? And also point out the other things which need to be added to merge this PR and next steps

sayakpaul commented:
@ayushtues could you please resolve the conversations if the comments were addressed? Since the trail is quite long, resolving them when appropriate really helps the reviewing process.

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion_controlnet.py:
```diff
@@ -0,0 +1,375 @@
+# Copyright 2023 Salesforce.com, inc.
```
Should we have it inside the `controlnet` module and rename it to `pipeline_controlnet_blip_diffusion.py`? Not too strongly opinionated about this either since BLIP Diffusion utilizes Stable Diffusion parameters anyway. 

@patrickvonplaten what are your thoughts?

patrickvonplaten replied:
Good point, I'd actually put it under the controlnet folder indeed

ayushtues replied:
Okay, shifting to controlnet module then

ayushtues replied:
Moved

sayakpaul commented on src/diffusers/image_processor.py:
```diff
@@ -13,7 +13,7 @@
 
 import warnings
-from typing import List, Optional, Union
+from typing import List, Optional, Union, Iterable
```
Where are we using `Iterable`?

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/blip_image_processing.py:
```diff
@@ -0,0 +1,456 @@
+# coding=utf-8
+# Copyright 2022 The HuggingFace Inc. team. All rights reserved.
```
nit: "2023".

For other reviewers, it would be useful if we could add a comment directly in the code explaining why we're adding a new module specifically for image processing that seems to contain a lot of copied-over functions from `transformers`.

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/blip_image_processing.py:
```diff
@@ -0,0 +1,456 @@
+        encoded_outputs = BatchFeature(data={"pixel_values": images}, tensor_type=return_tensors)
+        return encoded_outputs
+
+    def canny_preprocess(
```
We don't include separate functions for preprocessing condition images (as needed by models like ControlNet) from the `diffusers`. These preprocessing modules should be extremely lightweight in the sense that they should focus on the bare minium but important aspects of preprocessing (such as rescaling, extension conversions, etc.). But these kinds of preprocessing should remain with the users. Please refer to the official ControlNet documentation to understand how we deal with them. 

Of course, feel free to ask me questions if something is unclear.

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/blip_image_processing.py:
```diff
@@ -0,0 +1,456 @@
+            images = images.repeat_interleave(2, dim=0)
+        return images
+
+    # Follows diffusers.VaeImageProcessor.postprocess
```
If it's copied from the `VaeImageProcessor.postprocess()` let's add a proper "# Copied from ..." comment here.

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion_controlnet.py:
```diff
@@ -0,0 +1,375 @@
+
+class BlipDiffusionControlNetPipeline(DiffusionPipeline):
+    """
+    Pipeline for Canny Edge based Controlled subject-driven generation using Blip Diffusion.
```
If a user loads a ControlNet trained on say, depth conditions, I think we should be able to handle that here too, no? 

I think for that to work successfully, we need to delegate the preprocessing steps (for the conditioning image as needed by a ControlNet) to the user and then it should work right off the bat. This delegation is absolutely fine because we already do it for our existing ControlNets. 

What are your thoughts? 

This way, I think we open various doors of very interesting applications.

ayushtues replied:
Yeah makes sense, let me refactor and try to add 2 different controlnet examples

ayushtues replied:
Done

sayakpaul commented on tests/pipelines/blipdiffusion/test_blipdiffusion.py:
```diff
@@ -0,0 +1,178 @@
+import gc
```
Missing licensing. Also the name of the file should start with "test_pipeline".

ayushtues replied:
@sayakpaul I can't see the other test files following that convention, https://github.com/huggingface/diffusers/tree/main/tests/pipelines/kandinsky

sayakpaul replied:
Ah okay to ignore then. But the licensing is missing.

sayakpaul reviewed (commented):
Looking very nice! 

Left another round of comments. From my end the major points are:

* Discarding the ControlNet conditioning preprocessing from our codebase and leaving that to the users to take care of (as we already do that). 
* Use smaller models for fast tests. Ideally, these should run very quickly. The current ones are a bit slower. 
* We need to add an entry for BLIP Diffusion to our `docs`. 

Finally, I would like to involve @yiyixuxu to brainstorm how the BLIP image preprocessing utilities should be included in the PR. @yiyixuxu, for context, we need to have a bit of custom preprocessing for the images accepted by a BLIP diffusion pipeline since it leverages the BLIP image encoder module.

patrickvonplaten commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion_controlnet.py:
```diff
@@ -0,0 +1,406 @@
+    replace_example_docstring,
+)
+from ...utils.torch_utils import randn_tensor
+from .. import DiffusionPipeline
```
```suggestion
from ..pipeline_utils import DiffusionPipeline
```

patrickvonplaten commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,340 @@
+from transformers import CLIPTokenizer
+
+from ...models import AutoencoderKL, UNet2DConditionModel
+from ...pipelines import DiffusionPipeline
```
```suggestion
from ..pipeline_utils import DiffusionPipeline
```

patrickvonplaten commented on tests/pipelines/blipdiffusion/test_blipdiffusion.py:
```diff
@@ -0,0 +1,192 @@
+
+    def get_dummy_inputs(self, device, seed=0):
+        np.random.seed(seed)
+        reference_image = np.random.rand(64, 64, 3) * 255
```
Any chance we could test on smaller images? Our tests would be very thankful :sweat_smile:

ayushtues replied:
Changed to 32x32

patrickvonplaten commented on tests/pipelines/blipdiffusion/test_blipdiffusion_controlnet.py:
```diff
@@ -0,0 +1,212 @@
+        }
+
+        blip_qformer_config = {
+            "vocab_size": 30522,
```
Ufff that's quite big, this might lead to significant slow downs for CPU testing

ayushtues replied:
Changed to 1000

patrickvonplaten reviewed (approved):
Super clean PR, great job @ayushtues ! 

From my side this looks more or less good to merge. 

1. How long do the fast tests take here? Slightly worried about these tests increasing the time too much
2. Good to merge for me once @sayakpaul is happy with it

ayushtues commented:
Hey @sayakpaul, I refactored the Controlnet Processing, made the tests faster ( at least tried to ), and added some initial docs.  Might need some help with the docs, about what all needs to be in them.

sayakpaul commented on docs/source/en/api/pipelines/blip_diffusion.md:
```diff
@@ -0,0 +1,22 @@
+# Blip Diffusion
+
+Blip Diffusion was proposed in [BLIP-Diffusion: Pre-trained Subject Representation for Controllable Text-to-Image Generation and Editing
+](https://arxiv.org/abs/2305.14720) by Dongxu Li, Junnan Li, Steven C.H. Hoi
```
```suggestion
Blip Diffusion was proposed in [BLIP-Diffusion: Pre-trained Subject Representation for Controllable Text-to-Image Generation and Editing](https://arxiv.org/abs/2305.14720). It enables zero-shot subject-driven generation and control-guided zero-shot generation. 
```

sayakpaul replied:
WDYT?

sayakpaul commented on docs/source/en/api/pipelines/blip_diffusion.md:
```diff
@@ -0,0 +1,22 @@
+
+The abstract from the paper is:
+
+*Subject-driven text-to-image generation models create novel renditions of an input subject based on text prompts. Existing models suffer from lengthy fine-tuning and difficulties preserving the subject fidelity. To overcome these limitations, we introduce BLIP-Diffusion, a new subject-driven image generation model that supports multimodal control which consumes inputs of subject images and text  ...
```
```suggestion
*Subject-driven text-to-image generation models create novel renditions of an input subject based on text prompts. Existing models suffer from lengthy fine-tuning and difficulties preserving the subject fidelity. To overcome these limitations, we introduce BLIP-Diffusion, a new subject-driven image generation model that supports multimodal control which consumes inputs of subject images and text prompts. Unlike other subject-driven generation models, BLIP-Diffusion introduces a new multimodal encoder which is pre-trained to provide subject representation. We first pre-train the multimodal encoder following BLIP-2 to produce visual representation aligned with the text. Then we design a subject representation learning task which enables a diffusion model to leverage such visual representation and generates new subject renditions. Compared with previous methods such as DreamBooth, our model enables zero-shot subject-driven generation, and efficient fine-tuning for customized subject with up to 20x speedup. We also demonstrate that BLIP-Diffusion can be flexibly combined with existing techniques such as ControlNet and prompt-to-prompt to enable novel subject-driven generation and editing applications.*

sayakpaul commented on docs/source/en/api/pipelines/blip_diffusion.md:
```diff
@@ -0,0 +1,22 @@
+*Subject-driven text-to-image generation models create novel renditions of an input subject based on text prompts. Existing models suffer from lengthy fine-tuning and difficulties preserving the subject fidelity. To overcome these limitations, we introduce BLIP-Diffusion, a new subject-driven image generation model that supports multimodal control which consumes inputs of subject images and text  ...
+
+The original codebase can be found at [salesforce/LAVIS](https://github.com/salesforce/LAVIS/tree/main/projects/blip-diffusion).
+
```
Give yourself credit here, Ayush. Please add:

```
`BlipDiffusionPipeline` and `BlipDiffusionControlNetPipeline` were contributed by [`ayushtues`](https://github.com/ayushtues/).
```

sayakpaul commented on docs/source/en/api/pipelines/blip_diffusion.md:
```diff
@@ -0,0 +1,22 @@
+## BlipDiffusionControlNetPipeline
+[[autodoc]] BlipDiffusionControlNetPipeline
+    - all
+    - __call__
```
Do we need to add any output classes?

ayushtues replied:
We have the standard `ImagePipelineOutput` class, so I think not? Correct me if I am wrong.

sayakpaul replied:
If we used the standard output classes in the pipeline implementations, then yes, no need.

sayakpaul commented:
@ayushtues let's fix the CI here (don't worry about `build_pr_documentation` workflow for now). Then we can take a look at the test timings and reevaluate if we need to speed up the tests even more.

ayushtues commented:
Hey @sayakpaul is there a way to test the CI locally? Currently the workflows need approval, making it hard to debug

sayakpaul commented:
You can use `pytest ...` to run the corrected tests locally. But for repo consistency, it should `make fix-copies`. For styling, it should `make style && make quality`.

ayushtues commented:
> You can use `pytest ...` to run the corrected tests locally. But for repo consistency, it should `make fix-copies`. For styling, it should `make style && make quality`.

`make` working well now

sayakpaul commented:
Okay, let's see how it goes this time :)

charchit7 commented:
Nice work @ayushtues "**All Might**" will be happy.

sayakpaul commented:
@ayushtues there are still some open comments. Let's resolve them and we can then merge hopefully tomorrow.

sayakpaul commented:
Also, @patrickvonplaten I think we need to transfer the [weights](https://huggingface.co/ayushtues/blipdiffusion) to https://huggingface.co/Salesforce once the model cards are ready.

ayushtues commented:
The model's outputs seem a bit different from what they were in the beginning, similar quality, but different, will debug that first

ayushtues commented:
Hey @sayakpaul took a stab at creating the model card : https://huggingface.co/ayushtues/blipdiffusion/blob/main/README.md

sayakpaul commented:
Nice this works for me! Let's also make sure to change `ayushtues/blipdiffusion` to `SalesForce/blipdiffusion` before merging, though. This applies to the model card and the documentation. @patrickvonplaten could you please help us with the transfer? 

Let me play with the pipeline a bit as well and see I find anything off.

ayushtues commented:
Sure, also will need to transfer https://huggingface.co/ayushtues/blipdiffusion-controlnet, which has the controlnet pipeline

sayakpaul commented:
@ayushtues I took a deeper look at the examples and I think we still need some stuff in to maintain a consistent UX in terms of the APIs:

## `torch_dtype=torch.float16` should work but currently doesn't

```python
from diffusers.pipelines import BlipDiffusionPipeline
from diffusers.utils import load_image
import torch


blip_diffusion_pipe = BlipDiffusionPipeline.from_pretrained(
    "ayushtues/blipdiffusion", torch_dtype=torch.float16
).to("cuda")

cond_subject = "dog"
tgt_subject = "dog"
text_prompt_input = "swimming underwater"


cond_image = load_image(
    "https://huggingface.co/datasets/ayushtues/blipdiffusion_images/resolve/main/dog.jpg"
)
num_output = 1

iter_seed = 88888
guidance_scale = 7.5
num_inference_steps = 25
negative_prompt = "over-exposure, under-exposure, saturated, duplicate, out of frame, lowres, cropped, worst quality, low quality, jpeg artifacts, morbid, mutilated, out of frame, ugly, bad anatomy, bad proportions, deformed, blurry, duplicate"

output = blip_diffusion_pipe(
    text_prompt_input,
    cond_image,
    cond_subject,
    tgt_subject,
    guidance_scale=guidance_scale,
    num_inference_steps=num_inference_steps,
    neg_prompt=negative_prompt,
    height=512,
    width=512,
).images
output[0].save("image.png")
```

Leads to:

```
[/usr/local/lib/python3.10/dist-packages/torch/nn/functional.py](https://localhost:8080/#) in layer_norm(input, normalized_shape, weight, bias, eps)
   2513             layer_norm, (input, weight, bias), input, normalized_shape, weight=weight, bias=bias, eps=eps
   2514         )
-> 2515     return torch.layer_norm(input, normalized_shape, weight, bias, eps, torch.backends.cudnn.enabled)
   2516 
   2517 

RuntimeError: expected scalar type Float but found Half
```

Additionally, you'd notice that I am not passing lists for `cond_subject`, `tgt_subject`, and `text_prompt_input`. This is consistent w.r.t the other pipelines and we need to ensure that this stays the same for BLIP Diffusion, too. 

This applies to the ControlNet pipeline too. 

> Sure, also will need to transfer https://huggingface.co/ayushtues/blipdiffusion-controlnet, which has the controlnet pipeline

Then let's add a model card to that one too after resolving the above-mentioned issue.

ayushtues commented:
@sayakpaul fixed the fp16 issue and also changed examples not to use lists ( the pipeline worked fine without lists ). Also added model card for Controlnet

sayakpaul commented on docs/source/en/api/pipelines/blip_diffusion.md:
```diff
@@ -0,0 +1,23 @@
+
+*Subject-driven text-to-image generation models create novel renditions of an input subject based on text prompts. Existing models suffer from lengthy fine-tuning and difficulties preserving the subject fidelity. To overcome these limitations, we introduce BLIP-Diffusion, a new subject-driven image generation model that supports multimodal control which consumes inputs of subject images and text  ...
+
+The original codebase can be found at [salesforce/LAVIS](https://github.com/salesforce/LAVIS/tree/main/projects/blip-diffusion).
```
```suggestion
The original codebase can be found at [salesforce/LAVIS](https://github.com/salesforce/LAVIS/tree/main/projects/blip-diffusion). You can find the official BLIP Diffusion checkpoints under the [hf.co/SalesForce](https://hf.co/SalesForce) organization.
```

sayakpaul replied:
Let's also add a tips section. You can take a reference from this file: https://github.com/huggingface/diffusers/blob/main/docs/source/en/api/pipelines/controlnet.md (view the source and look for "Make sure to check out the Schedulers ...").

sayakpaul commented on docs/source/en/api/pipelines/blip_diffusion.md:
```diff
@@ -0,0 +1,23 @@
+## BlipDiffusionControlNetPipeline
+[[autodoc]] BlipDiffusionControlNetPipeline
+    - all
+    - __call__
```
We also need to add an entry about the doc to https://github.com/huggingface/diffusers/blob/main/docs/source/en/_toctree.yml.

sayakpaul commented on src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py:
```diff
@@ -0,0 +1,338 @@
+            # to avoid doing two forward passes
+            text_embeddings = torch.cat([uncond_embeddings, text_embeddings])
+
+        scale_down_factor = 2 ** (len(self.unet.block_out_channels) - 1)
```
Let's access the model configuration variables from the `config` argument i.e., `self.unet.config.block_out_channels`. Applies to wherever applicable.

sayakpaul commented:
@ayushtues hopefully final set of comments from my end before we can merge:

* Resolve the open comments. 
* Change the examples to reflect the commonly followed practices:

**Zero-shot:**

```python
from diffusers.pipelines import BlipDiffusionPipeline
from diffusers.utils import load_image
import torch

blip_diffusion_pipe = BlipDiffusionPipeline.from_pretrained(
    "ayushtues/blipdiffusion", torch_dtype=torch.float16
).to("cuda")

cond_subject = "dog"
tgt_subject = "dog"
text_prompt_input = "swimming underwater"

cond_image = load_image(
    "https://huggingface.co/datasets/ayushtues/blipdiffusion_images/resolve/main/dog.jpg"
)

iter_seed = 88888
guidance_scale = 7.5
num_inference_steps = 25
negative_prompt = "over-exposure, under-exposure, saturated, duplicate, out of frame, lowres, cropped, worst quality, low quality, jpeg artifacts, morbid, mutilated, out of frame, ugly, bad anatomy, bad proportions, deformed, blurry, duplicate"

output = blip_diffusion_pipe(
    text_prompt_input,
    cond_image,
    cond_subject,
    tgt_subject,
    guidance_scale=guidance_scale,
    num_inference_steps=num_inference_steps,
    neg_prompt=negative_prompt,
    height=512,
    width=512,
).images
output[0].save("image.png")
```

**Control-guided (Canny):**

```python
from diffusers.pipelines import BlipDiffusionControlNetPipeline
from diffusers.utils import load_image
from controlnet_aux import CannyDetector

blip_diffusion_pipe = BlipDiffusionControlNetPipeline.from_pretrained(
    "ayushtues/blipdiffusion-controlnet", torch_dtype=torch.float16
).to("cuda")

style_subject = "flower"  # subject that defines the style
tgt_subject = "teapot"  # subject to generate.
text_prompt = "on a marble table"

cldm_cond_image = load_image(
    "https://huggingface.co/datasets/ayushtues/blipdiffusion_images/resolve/main/kettle.jpg"
).resize((512, 512))
canny = CannyDetector()
cldm_cond_image = canny(cldm_cond_image, 30, 70, output_type="pil")
style_image = load_image(
    "https://huggingface.co/datasets/ayushtues/blipdiffusion_images/resolve/main/flower.jpg"
)

guidance_scale = 7.5
num_inference_steps = 50
negative_prompt = "over-exposure, under-exposure, saturated, duplicate, out of frame, lowres, cropped, worst quality, low quality, jpeg artifacts, morbid, mutilated, out of frame, ugly, bad anatomy, bad proportions, deformed, blurry, duplicate"

output = blip_diffusion_pipe(
    text_prompt,
    style_image,
    cldm_cond_image,
    style_subject,
    tgt_subject,
    guidance_scale=guidance_scale,
    num_inference_steps=num_inference_steps,
    neg_prompt=negative_prompt,
    height=512,
    width=512,
).images
output[0].save("image.png")
```

**Control-guided (scribble):**

```python
from diffusers.pipelines import BlipDiffusionControlNetPipeline
from diffusers.utils import load_image
from controlnet_aux import HEDdetector

blip_diffusion_pipe = BlipDiffusionControlNetPipeline.from_pretrained(
    "ayushtues/blipdiffusion-controlnet"
)
controlnet = ControlNetModel.from_pretrained("lllyasviel/sd-controlnet-scribble")
blip_diffusion_pipe.controlnet = controlnet
blip_diffusion_pipe.to("cuda")

style_subject = "flower"  # subject that defines the style
tgt_subject = "bag"  # subject to generate.
text_prompt = "on a table"
cldm_cond_image = load_image(
    "https://huggingface.co/lllyasviel/sd-controlnet-scribble/resolve/main/images/bag.png"
).resize((512, 512))
hed = HEDdetector.from_pretrained("lllyasviel/Annotators")
cldm_cond_image = hed(cldm_cond_image)
style_image = load_image(
    "https://huggingface.co/datasets/ayushtues/blipdiffusion_images/resolve/main/flower.jpg"
)

guidance_scale = 7.5
num_inference_steps = 50
negative_prompt = "over-exposure, under-exposure, saturated, duplicate, out of frame, lowres, cropped, worst quality, low quality, jpeg artifacts, morbid, mutilated, out of frame, ugly, bad anatomy, bad proportions, deformed, blurry, duplicate"

output = blip_diffusion_pipe(
    text_prompt,
    style_image,
    cldm_cond_image,
    style_subject,
    tgt_subject,
    guidance_scale=guidance_scale,
    num_inference_steps=num_inference_steps,
    neg_prompt=negative_prompt,
    height=512,
    width=512,
).images
output[0].save("image.png")
```

Would be great if you could update the documentation to reflect this and also the model cards. 

Then we can edit the checkpoint path to reflect `SalesForce`, transfer them, and finally merge the PR. 

Let me know if anything's unclear.

sayakpaul commented:
Another point is have you tried using the pipeline on various subjects and seeing if it's able to faithfully render them in the outputs? 

For example, I tried the zero-shot rendition pipeline on [this image](https://datasets-server.huggingface.co/assets/google/dreambooth/--/backpack/train/2/image/image.jpg) with the following parameters:

```
cond_subject = "backpack"
tgt_subject = "backpack"
text_prompt_input = "in a busy street"
```

But it didn't faithfully render the subject:

![image](https://github.com/huggingface/diffusers/assets/22957388/986c11e1-60aa-4069-97ec-b8eb6ad4dbb7)


Is this expected?

ayushtues commented:
![image](https://github.com/huggingface/diffusers/assets/43698245/1337433f-e960-4adf-9823-f939a2e7884c)
![image](https://github.com/huggingface/diffusers/assets/43698245/be853c71-10c8-4c37-b2fc-121d9e8b355f)

These are some examples I seem to be getting

dxli94 commented:
Thanks @ayushtues @sayakpaul for the addition. The results look as expected. In zero-shot inference, the subject appearance does deviate a bit as suggested in the reported metrics (CLIP-I, DINO). More similar results can be obtained by few-step fine-tuning.

ayushtues commented:
^ @sayakpaul, if the original author says so :P

patrickvonplaten reviewed (approved):
Good to merge from my side

sayakpaul commented:
Alright. Really, great work here Ayush!

We have asked internally to move the checkpoints to hf.co/Salesforce. Once this is done and the checkpoint paths have been updated in the model cards (should be done before we transfer actually) and the docs, I will merge the PR :)

ayushtues commented:
> Alright. Really, great work here Ayush!
> 
> We have asked internally to move the checkpoints to hf.co/Salesforce. Once this is done and the checkpoint paths have been updated in the model cards (should be done before we transfer actually) and the docs, I will merge the PR :)

Let me know when the transfer is done, I'll do the other changes

sayakpaul commented:
@ayushtues we have got approval from @dxli94 to do the transfer. Please update the checkpoint paths and once done, let me know here. Will transfer and merge.

ayushtues commented:
Hey @sayakpaul  updated to `Salesforce/blipdiffusion` & `Salesforce/blipdiffusion-controlnet`, let me know if anything else is needed

sayakpaul reviewed (approved):
Amazing work! Thanks so much for your patience and for iterating!

As soon as the transfer is done, will merge and ship this 🚀

sayakpaul commented:
Trasfer complete! Merging!

yanchaoguo commented:
> Trasfer complete! Merging!

Excellent! I have used it
![企业微信截图_16953670298817](https://github.com/huggingface/diffusers/assets/11827802/4ed5ae7e-0af5-438d-84e3-a80793b23295)
![image](https://github.com/huggingface/diffusers/assets/11827802/f5d4ac8d-a79c-4eb7-ba55-8e3b207ef847)

yanchaoguo commented:
how to load lora file using blip diffusion @sayakpaul
